app/code/Magento/InventorySales/Model/RegisterSalesEvent.php

```diff
@@ -0,0 +1,137 @@
+        GetSkusByProductIdsInterface $getSkusByProductIds,
+        StockByWebsiteIdResolver $stockByWebsiteIdResolver,
+        ReservationBuilderInterface $reservationBuilder,
+        AppendReservationsInterface $appendReservations,
+        IsProductSalableForRequestedQtyInterface $isProductSalableForRequestedQty,
+        StockResolverInterface $stockResolver,
+        IsSourceItemsAllowedForProductTypeInterface $isSourceItemsAllowedForProductType,
+        GetProductTypesBySkusInterface $getProductTypesBySkus
+    ) {
+        $this->getSkusByProductIds = $getSkusByProductIds;
+        $this->stockByWebsiteIdResolver = $stockByWebsiteIdResolver;
+        $this->reservationBuilder = $reservationBuilder;
+        $this->appendReservations = $appendReservations;
+        $this->isProductSalableForRequestedQty = $isProductSalableForRequestedQty;
+        $this->stockResolver = $stockResolver;
+        $this->isSourceItemsAllowedForProductType = $isSourceItemsAllowedForProductType;
+        $this->getProductTypesBySkus = $getProductTypesBySkus;
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function execute(array $items, SalesChannelInterface $salesChannel, SalesEventInterface $salesEvent)
+    {
+        if (empty($items)) {
+            return;
+        }
+
+        if (null === $salesChannel) {
+            throw new LocalizedException(__('$salesChannel parameter is required'));
+        }
+
+        // TODO typecast needed because StockInterface::getStockId() returns string => fix StockInterface::getStockId?
+        $stockId = (int)$this->stockResolver->get($salesChannel->getType(), $salesChannel->getCode())->getStockId();
+
+        $productTypes = $this->getProductTypesBySkus->execute(array_keys($items));
+        $this->checkItemsQuantity($items, $productTypes, $stockId);
+        $reservations = [];
+        foreach ($items as $sku => $qty) {
+            $reservations[] = $this->reservationBuilder
+                ->setSku($sku)
+                ->setQuantity(-(float) $qty)
+                ->setStockId($stockId)
+                ->setMetadata(sprintf('%s:%d', $salesEvent->getType(), $salesEvent->getObjectId()))
+                ->build();
+        }
+        $this->appendReservations->execute($reservations);
+    }
+
+    /**
+     * Check is all items salable
+     *
+     * @return void
+     * @throws LocalizedException
+     */
+    private function checkItemsQuantity(array $items, array $productTypes, int $stockId)
+    {
+        foreach ($items as $sku => $qty) {
+            if (false === $this->isSourceItemsAllowedForProductType->execute($productTypes[$sku])) {
+                continue;
+            }
+            $isSalable = $this->isProductSalableForRequestedQty->execute($sku, $stockId, $qty)->isSalable();
+            if (false === $isSalable) {
+                throw new LocalizedException(
+                    __('Not all of your products are available in the requested quantity.')
+                );
```

Comment:
this code 
```
            $isSalable = $this->isProductSalableForRequestedQty->execute($sku, $stockId, $qty)->isSalable();
            if (false === $isSalable) {
                throw new LocalizedException(
                    __('Not all of your products are available in the requested quantity.')
                );
            }
```

in incorrect, first of all, because `isProductSalableForRequestedQty->execute` returns object but not boolean.
And the object returned `\Magento\InventorySalesApi\Api\Data\ProductSalableResultInterface`
already contains high-level error message, so no need to overwrite it with standard `'Not all of your products are available in the requested quantity.'` because now you can lose messages about Qty Increment or similar.